fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/Binder.java

```diff
@@ -162,6 +153,108 @@ private GroupExpression extractGroupExpression(Pattern pattern, Group group) {
+     */
+    class MultiJoinBinder {
+
+        public OptExpression match(GroupExpression ge) {
+            // 1. Check if the entire tree is MULTI_JOIN
+            // 2. Enumerate GE
+            if (ge == null || !isMultiJoin(ge)) {
+                return null;
+            }
+
+            return enumerate(ge);
+        }
+
+        private OptExpression enumerate(GroupExpression ge) {
+            if (ge == null || !isMultiJoinGE(ge)) {
+                return null;
+            }
+            List<OptExpression> resultInputs = Lists.newArrayList();
+
+            int groupExpressionIndex = 0;
+            while (groupExpressionIndex < ge.getInputs().size()) {
+                trace();
+                Group group = ge.getInputs().get(groupExpressionIndex);
+                OptExpression opt = enumerate(extractGroupExpression(group));
+
+                if (opt == null) {
+                    return null;
+                } else {
+                    resultInputs.add(opt);
+                }
+
+                groupExpressionIndex++;
+            }
+
+            return new OptExpression(ge, resultInputs);
+        }
+
+        private GroupExpression extractGroupExpression(Group group) {
+            int valueIndex = groupExpressionIndex.get(groupTraceKey);
+            if (valueIndex >= group.getLogicalExpressions().size()) {
+                groupExpressionIndex.remove(groupTraceKey);
+                return null;
+            }
+            return group.getLogicalExpressions().get(valueIndex);
+        }
+
+        private boolean isMultiJoin(GroupExpression ge) {
+            return ge.getOp().getOpType() == OperatorType.LOGICAL_JOIN && isMultiJoinRecursive(ge);
+        }
+
+        /**
+         * Recursive check whether the GE is a MULTI_JOIN
+         * For nested-mv, the original GE may be a AGG, after rewriting this group could be put an SCAN, so
+         * we also need to consider this kind of GE as MULTI_JOIN
+         */
+        private boolean isMultiJoinRecursive(GroupExpression ge) {
+            if (!isMultiJoinOp(ge.getOp().getOpType())) {
+                return false;
+            }
+
+            for (int i = 0; i < ge.getInputs().size(); i++) {
+                Group child = ge.inputAt(i);
+                if (isMultiJoinRecursive(child.getFirstLogicalExpression())) {
+                    continue;
+                }
+                if (!hasRewrittenMvScan(child)) {
+                    return false;
+                }
+            }
+            return true;
+        }
+
+        private boolean hasRewrittenMvScan(Group g) {
+            for (GroupExpression ge : g.getLogicalExpressions()) {
+                if ((isScanOp(ge.getOp().getOpType()))) {
+                    return true;
+                }
+            }
+            return false;
+        }
+
+        private boolean isScanOp(OperatorType operatorType) {
+            return Pattern.ALL_SCAN_TYPES.contains(operatorType);
```

Comment:
MV only can be OLAP_SCAN_TYPE?

Reply:
good point, it could prune the useless search space further.

Reply:
fixed